File: PDOO/Napakalaki-Ruby/lib/Napakalaki.rb
#encoding: utf-8

require 'singleton'

require_relative 'Player.rb'
require_relative 'CardDealer.rb'
require_relative 'CombatResult.rb'
require_relative 'Monster.rb'
require_relative 'CultistPlayer.rb'

module NapakalakiGame
    
class Napakalaki
    
    include Singleton
    
    attr_accessor :currentPlayer, :players, :currentMonster, :dealer, :currentPlayerIndex
    
    def initialize
        @currentPlayer=nil
        @currentPlayerIndex=-1
        @players=Array.new
        @dealer=CardDealer.instance
        @currentMonster=nil
    end
    
    private
    
    def initPlayers(names)
        names.each do |name|
            @players << Player.new(name)
        end
    end
    
    def nextPlayer
        if(@currentPlayerIndex==-1)
            rnd=rand(@players.size())
            @currentPlayer=@players[rnd]
            @currentPlayerIndex=rnd
        else
            @currentPlayerIndex=@currentPlayerIndex+1
            if(@currentPlayerIndex==@players.size())
                @currentPlayerIndex=0
            end
        end
        
        @currentPlayer=@players[@currentPlayerIndex]
        return @currentPlayer
    end
    
    def nextTurnIsAllowed
        if(@currentPlayer==nil)
            return true
        else
            return @currentPlayer.validState
        end
    end
    
    def setEnemies
        for i in(0..@players.size() -1)
            n = rand(@players.size())
            while (i == n)
                n = rand(@players.size())
            end
            @players[i].setEnemy(@players[n])
        end
    end
    
    #------------------------ PARTE PÚBLICA  ------------------------
    
    
    public
    
    def developCombat()
      combatresult = @currentPlayer.combat(@currentMonster)
      @dealer.giveMonsterBack(@currentMonster)
      if(combatresult == CombatResult::LOSEANDCONVERT)
        aux = CultistPlayer.new(@currentPlayer,@dealer.nextCultist)
        @players.delete(@currentPlayer)
        @currentPlayer = aux
        @players << aux
      end
      return combatresult
    end
    
    def discardVisibleTreasures(treasures)
        for i in 0..(treasures.size - 1)
          treasure = treasures[i]
          @currentPlayer.discardVisibleTreasure(treasure)
          @dealer.giveTreasureBack(treasure)
        end
    end
    
    def discardHiddenTreasures(treasures)
        for i in 0..(treasures.size - 1)
          treasure = treasures[i]
          @currentPlayer.discardHiddenTreasure(treasure)
          @dealer.giveTreasureBack(treasure)
        end
    end
    
    def makeTreasuresVisible(treasure)
        treasure.each do | t |
          @currentPlayer.makeTreasureVisible(t)
        end
    end
    
    def initGame(names)
        initPlayers(names)
        setEnemies();
        dealer.initCards()
        nextTurn()
    end
    
    def getCurrentPlayer
        return @currentPlayer
    end
    
    def getCurrentMonster
        return @currentMonster
    end
    
    def nextTurn
        stateOK = nextTurnIsAllowed
        if(stateOK)
          @currentMonster =@dealer.nextMonster
          @currentPlayer = nextPlayer
          dead = @currentPlayer.isDead
          if(dead)
            @currentPlayer.initTreasures
          end
        end
        return stateOK
    end
    
    
    def endOfGame(result)
        if(result == CombatResult::WINGAME)
            return true
        else
            return false
        end
    end
    
end
end

File: PDOO/Napakalaki-Java/src/NapakalakiGame/Napakalaki.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package NapakalakiGame;

import java.util.*;

/**
 *
 * @author Usuario
 */
public class Napakalaki {
    
    private static Napakalaki instance = null;
    
    private Player currentPlayer;
    private int currentPlayerIndex;
    private ArrayList<Player> players;
    private CardDealer dealer;
    private Monster currentMonster;
    
    public static Napakalaki getInstance(){
        if(instance==null)
            instance=new Napakalaki();
        return instance;
    }
    
    private Napakalaki(){
        currentPlayer=null;
        currentPlayerIndex=-1;
        players=new ArrayList();
        dealer=CardDealer.getInstance();
        currentMonster=null;
    }
    
    public int getCurrenPlayerIndex(){
        return currentPlayerIndex;
    }
    
    public ArrayList<Player> getPlayers(){
        return players;
    }
    
    private void initPlayers(ArrayList<String> names){
        for(int i=0; i<names.size(); i++){
            String name = names.get(i);
            Player player = new Player(name);
            players.add(player);
        }
    }
    
    private Player nextPlayer(){
        if(currentPlayerIndex==-1){     //Primera jugada
            Random rnd = new Random();
            currentPlayerIndex=(int)(rnd.nextDouble() * players.size());
        }
        else{
            currentPlayerIndex++;
            if(currentPlayerIndex==players.size())
                currentPlayerIndex=0;
        }
        
        currentPlayer=players.get(currentPlayerIndex);
        return currentPlayer;
     }
    
    private boolean nextTurnIsAllowed(){
        if(currentPlayer==null)
            return true;
        
        return currentPlayer.validState();
    }
    
    private void setEnemies(){
        Random rnd = new Random();
        
        for (int i=0; i <players.size(); i++){
            int randomPos = rnd.nextInt(players.size());
            
            while (i == randomPos)
                randomPos = rnd.nextInt(players.size());
            
            players.get(i).setEnemy(players.get(randomPos));
        }
    }
    
   public CombatResult developCombat(){
        Monster m = currentMonster;
        CombatResult combatResult = currentPlayer.combat(m);
        if(combatResult == CombatResult.LOSEANDCONVERT){
            Cultist c = dealer.nextCultist();
            Player p = new CultistPlayer(currentPlayer, c);
            players.add(p);
            players.remove(currentPlayer);
            currentPlayer = p;
        }            
        dealer.giveMonsterBack(m);
        return combatResult;
    }
    
    public void discardVisibleTreasures(ArrayList<Treasure> treasures){
        for(int i=0; i<treasures.size(); i++){
            Treasure treasure = treasures.get(i);
            currentPlayer.discardVisibleTreasure(treasure);
            dealer.giveTreasureBack(treasure);
        }
    }
    
    public void discardHiddenTreasures(ArrayList<Treasure> treasures){
        for(int i=0; i<treasures.size(); i++){
            Treasure treasure = treasures.get(i);
            currentPlayer.discardHiddenTreasure(treasure);
            dealer.giveTreasureBack(treasure);
        }
    }
    
    public void makeTreasuresVisible(ArrayList<Treasure> treasures){
        for(int i=0; i<treasures.size(); i++){
            Treasure t = treasures.get(i);
            currentPlayer.makeTreasureVisible(t);
        }
    }
    
    public void initGame(ArrayList<String> names){
        this.initPlayers(names);
        this.setEnemies();
        dealer.initCards();
        this.nextTurn();
    }
    
    public Player getCurrentPlayer(){
        return currentPlayer;
    }
    
    public Monster getCurrentMonster(){
        return currentMonster;
    }
    
    public boolean nextTurn(){
        boolean stateOK=this.nextTurnIsAllowed();
        if(stateOK){
            currentMonster = dealer.nextMonster();
            currentPlayer = this.nextPlayer();
            boolean dead = currentPlayer.isDead();
            if(dead){
                currentPlayer.initTreasures();
            }
        }
        return stateOK;
    }
    
    public boolean endOfGame(CombatResult result){
        if(result == CombatResult.WINGAME)
            return true;
        return false;
    }
}
